tutorials/detailed_model_reference.ipynb

```diff
@@ -0,0 +1,663 @@
+    "            reference_multivector = scatter(x, batch, dim=0, reduce='mean').mean(dim=1, keepdim=True)[batch]\n",
+    "\n",
+    "        return reference_multivector\n",
+    "\n",
+    "    def lift(self, multivectors: torch.Tensor, scalars: torch.Tensor) -> torch.Tensor:\n",
+    "\n",
+    "        if multivectors.size(0) == self.cache['data'].scale0_sampling_index.numel():\n",
+    "            multivectors, scalars = interp(\n",
+    "                self.mlp,\n",
+    "                multivectors,\n",
+    "                self.cache['multivectors'],\n",
+    "                scalars,\n",
+    "                self.cache['scalars'],\n",
+    "                self.cache['pos'],\n",
+    "                self.cache['data'].pos,\n",
+    "                self.cache['data'],\n",
+    "                scale_id=0,\n",
+    "                reference_multivector=self.cache['reference_multivector']\n",
+    "            )\n",
+    "\n",
+    "        else:\n",
+    "            multivectors, scalars = self.extract_class(\n",
+    "                self.mlp,\n",
+    "                multivectors,\n",
+    "                self.cache['multivectors'],\n",
+    "                scalars,\n",
+    "                self.cache['scalars'],\n",
+    "                self.cache['data']\n",
+    "            )\n",
+    "\n",
+    "        return self.geometric_algebra_interface.dislodge(multivectors, scalars)\n",
+    "\n",
+    "    def extract_class(\n",
+    "        self,\n",
+    "        mlp: MLP,\n",
+    "        multivectors: torch.Tensor,\n",
+    "        multivectors_skip: torch.Tensor,\n",
+    "        scalars: torch.Tensor,\n",
+    "        scalars_skip: torch.Tensor,\n",
+    "        data: Data\n",
+    "    ) -> torch.Tensor:\n",
+    "\n",
+    "        if data.batch is None:\n",
+    "            multivectors_skip = multivectors_skip.mean(dim=0, keepdim=True)\n",
+    "            scalars_skip = scalars_skip.mean(dim=0, keepdim=True)\n",
+    "\n",
+    "            reference_multivector = self.cache['reference_multivector'][0:1]\n",
+    "\n",
+    "        else:\n",
+    "            multivectors_skip = scatter(multivectors_skip, data.batch, dim=0, reduce='mean')\n",
+    "            scalars_skip = scatter(scalars_skip, data.batch, dim=0, reduce='mean')\n",
+    "\n",
+    "            reference_multivector = self.cache['reference_multivector'][data.ptr[:-1]]\n",
+    "\n",
+    "        multivectors = torch.cat((multivectors, multivectors_skip), dim=-2)\n",
+    "        scalars = torch.cat((scalars, scalars_skip), dim=-1)\n",
+    "\n",
+    "        return mlp(multivectors, scalars, reference_mv=reference_multivector)\n",
+    "\n"
+   ],
+   "id": "c2083f9c4b8a1833",
+   "outputs": [],
+   "execution_count": 8
+  },
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": "### 2.2 LaB-GATr",
+   "id": "d8a241a8fdd731e6"
+  },
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": "```TODO```",
```

Comment:
Now we can assemble the complete LaB-GATr model. It is comprised of tokenization and a GATr backend. For correct handling of PyG-style batching, we use block-diagonal attention masks. Another highlight is the class token support which we implement via a `class_token_forward_wrapper`.